Fix missing / unavailable tracks in search results

kingosticks opened the pull request:
This fixes #97 by setting the market parameter when doing searches using the Web API.

We don't currently set the parameter and it defaults to 'US' (https://github.com/spotify/web-api/issues/194). Depending on the logged in user's country, this may result in missing or unavailable tracks being returned. For example "The black box revelation" as reported in https://github.com/mopidy/mopidy-spotify/issues/106.

I appreciate that considering https://github.com/mopidy/mopidy-spotify/issues/130, this might not see the light of day, but at least this should remind us to set it as part of those future changes.

trygveaa commented on mopidy_spotify/search.py:
```diff
@@ -55,11 +55,18 @@ def search(config, session, requests_session,
+
+    try:
+        params['market'] = session.user_country
+    except ValueError as exc:
```
Can this ever happen?

kingosticks replied:
I don't really know, but remember, this is libspotify....

kingosticks replied:
I guess if we were really worried about libspotify returning non utf-8 bytes back to us we should handle it in pyspotify. I can remove this (and it's test).

trygveaa replied:
Hm, in libspotify it is an int and pyspotify doesn't raise any errors no matter what the value is, so I don't see how anything could happen. If the int is 0, you will get null-bytes, but that shouldn't trigger a ValueError.

kingosticks replied:
If that returned int is greater than 65535, then `utils.to_country` will raise ValueError when it tries to convert a number greater 255 to `chr`. That routine could do with another `& 0xff` in there.

kingosticks commented:
Squashed and I've also removed the exception handling from this PR.  It'd only be needed if libspotify started misbehaving (which we have zero reports of), in which case it should be handled in pyspotify rather than here.

jodal commented:
Is this necessary now when all requests are authenticated? Couldn't Spotify know from your credentials which market you're in?

kingosticks commented:
You'd hope so, but not quite. From https://developer.spotify.com/web-api/search-item/:

> If from_token is given and a valid access token is supplied in the request header, only items with content playable in the country associated with the user's account will be returned. (The country associated with the user's account can be viewed by the user in their account settings at https://www.spotify.com/se/account/overview/). Note that the user must have granted access to the user-read-private scope when the access token was issued.

If you don't specify anything, you used to default to US but I'm not sure if that's still true.

adamcik commented:
> If market=from_token is supplied in the query string, the current user must have authorized the user-read-private scope.

So yeah, we do need the scope even if we never directly look it up.

jodal commented:
And @kingosticks needs to update this PR to apply to develop again :-)

kingosticks commented:
Well that didn't work

kingosticks commented:
That's better!
